### PR TITLE
Fix ReferenceList

### PR DESCRIPTION
**Pull request by Murmele**
There were an issue that if only a remote was available, it shows the id instead of the remote name.

So I started to refactoring to make it more clear and easier to read. And writing unittests to not make it a third time wrong ....

### CODE REVIEW

**Review by exactly-one-kas — APPROVED** [review]

File: src/ui/ReferenceView.cpp [hunk 1]
```diff
@@ -43,7 +43,7 @@ bool refComparator(const git::Reference &lhs, const git::Reference &rhs) {
           lhsCommit.committer().date() > rhsCommit.committer().date());
 }
 
-enum ReferenceType { Branches = 0, Remotes = 1, Tags = 2 };
+enum ReferenceType { HEADER = 0, Branches = 1, Remotes = 2, Tags = 3 };
```
[the remark this answers is not in the document]
Comment (Murmele) [reply src/ui/ReferenceView.cpp:46]: No no in that case it is the header of the combobox:
image
Do you have any recomandation to rename it somehow?

[the remark this answers is not in the document]
Comment (exactly-one-kas) [reply src/ui/ReferenceView.cpp:46]: Ahh, then I'd say `COMBOBOX_HEADER` to clarify

[the remark this answers is not in the document]
Comment (Murmele) [reply src/ui/ReferenceView.cpp:46]: Perfect thanks I will change

File: src/ui/ReferenceView.cpp [hunk 2]
```diff
@@ -193,33 +211,37 @@ class ReferenceModel : public QAbstractItemModel {
...
-      return 0;
+    if (parent.internalId() != HEADER)
+      return 0; // refs it self do not have childs
 
-    return mRefs.at(parent.row()).refs.size();
+    if (parent.row() < mRefs.size()) {
+      return mRefs.at(parent.row()).refs.size();
+    } else {
+      return 0;
+    }
```
Comment (Murmele) [inline src/ui/ReferenceView.cpp:244]: Did you find a way to get into this problem or is it as a general prevention?

File: src/ui/ReferenceView.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Murmele) [reply src/ui/ReferenceView.cpp]: changed in d7de06cbb07d10a2ee877adbf4658a62ffcca8be

File: src/ui/ReferenceView.cpp (hunk 2 above)
Comment (exactly-one-kas) [reply src/ui/ReferenceView.cpp:244]: The CI test failed on Linux and this was reproducible on my machine as well

But I have no idea why it happens at all, but this fixes it by not throwing ¯\_(ツ)_/¯

Comment (Murmele) [reply src/ui/ReferenceView.cpp:244]: I will check out why it fails, because on my linux it did not fail

Comment (exactly-one-kas) [reply src/ui/ReferenceView.cpp:244]: I simply let the test run locally via `ninja check`

**Comment by Murmele** [comment]
Ok the problem in the unittest is, that branches are hidden and therefore the index is wrong (Remotes Index is hardcoded)

**Comment by Murmele** [comment]
@exactly-one-kas I found the issue. I created unittests for it so we don't need your workaround. I reverted the changes in the branches_panel test. I have them stored locally so if we need at some point we can use them

**Comment by Murmele** [comment]
image

Shall we use here really AllRefs? Because then Branches and tags are also visible. If you select a branch or tag the newBranch button is disabled.
Is there a benefit of seening the other branches?

**Comment by exactly-one-kas** [comment]
In the merge dialog I'd like to see remote branches and tags as well since it affects the merge commit message
For the new branch dialog I don't know why there even is the option to choose the reference since the "source reference" won't be mentioned anywhere. So I'd say we could just remove the control there

**Comment by Murmele** [comment]
> In the merge dialog I'd like to see remote branches and tags as well since it affects the merge commit message For the new branch dialog I don't know why there even is the option to choose the reference since the "source reference" won't be mentioned anywhere. So I'd say we could just remove the control there

Sorry my fault. It is the startpoint for NewBranch. This is fine. So on my point I don't have anything else. So from my side you can merge

**Comment by Murmele** [comment]
@exactly-one-kas can we merge this?

**Comment by exactly-one-kas** [comment]
yep

**Pull request merged**